bento_drop_box_service/backends/s3.py

```diff
@@ -0,0 +1,141 @@
+                    # If the directory is not in the tree, create it
+                    directory_path = "/".join(directories[: i + 1])
+                    new_tree_node = DropBoxEntry(
+                        name=directory_name, filePath=directory_path, relativePath=directory_path, contents=[]
+                    )
+                    current_level.append(new_tree_node)
+                    current_level = new_tree_node["contents"]
+            # Add file to the tree, at the right place (current level)
+            new_file = DropBoxEntry(
+                name=file["name"],
+                filePath=file["filePath"],
+                relativePath="/" + file["relativePath"],
+                size=file.get("size"),
+                lastModified=file["lastModified"],
+                lastMetadataChange=file["lastMetadataChange"],
+                uri=file["uri"],
+            )
+            current_level.append(new_file)
+        return tree
+
+    async def get_directory_tree(self, sub_path: str | None = None) -> tuple[DropBoxEntry, ...]:
+        prefix = sub_path if sub_path else ""
+        async with await self._create_s3_client() as s3_client:
+            paginator = s3_client.get_paginator("list_objects_v2")
+            page_iterator = paginator.paginate(Bucket=self.bucket_name, Prefix=prefix)
+
+            files_list: list[DropBoxEntry] = []
+            async for page in page_iterator:
+                if "Contents" in page:
+                    for obj in page["Contents"]:
+                        key = obj["Key"]
+                        if key.count("/") < self.config.traversal_limit:
+                            entry = {
+                                "name": key.split("/")[-1],
+                                "filePath": key,
+                                "relativePath": key,
+                                "size": obj["Size"],
+                                "lastModified": obj["LastModified"].timestamp(),
+                                "lastMetadataChange": obj["LastModified"].timestamp(),
+                                "uri": f"{self.config.service_url_base_path}/objects/{key}",
+                            }
+                            files_list.append(entry)
+
+        return tuple(self.create_directory_tree(files_list))
+
+    async def upload_to_path(self, request: Request, path: str, content_length: int) -> Response:
+        path = secure_filename(path)
+        async with await self._create_s3_client() as s3_client:
+            await s3_client.put_object(Bucket=self.bucket_name, Key=path, Body=await request.body())
+
+        return Response(status_code=status.HTTP_204_NO_CONTENT)
+
+    async def _retrive_headers(self, path: str) -> dict[str, str]:
+        file_name = path.split("/")[-1]
+        async with await self._create_s3_client() as s3:
+            head = await s3.head_object(Bucket=self.bucket_name, Key=path)
+            content_length = head["ContentLength"]
```

Comment:
these 4 variables (below `head`) don't need to be defined inside the context manager (and I'm not sure they need to be defined at all, since they're just immediately used below)